### PR TITLE
Array comprehension

### PR DESCRIPTION
**Pull request by ninegua**
This PR provides the ability to inline `numpy.array` call into list comprehension to eliminate the intermediate list during array initialization.

For example:

```
>>> import numpy as np
>>> np.array([ i for i in range(10) ])
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> np.array([ [ i + j for j in range(4) ] for i in range(3) ])
array([[0, 1, 2, 3],
       [1, 2, 3, 4],
       [2, 3, 4, 5]])
```

Numba can now compile both cases in nopython mode. The nested array case wasn't previously possible even with the list comprehension support due to nested objects. With array comprehension, a 2D array is created and initialized without creating nested objects.

Another reason to advocate this support is that although comprehension used for initialization can usually be imitated by lifting scalar operations to array level in combination with `numpy.arange`, this is not always possible or efficient when there is conditionals, as partly mentioned in #2168

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Also, please could the documentation be updated to reflect this change? Thanks.

File: numba/inline_closurecall.py [hunk 1]
```diff
@@ -1,11 +1,18 @@
 from numba import config, ir, ir_utils, utils
 import types
 
-from numba.ir_utils import (mk_unique_var, next_label, add_offset_to_labels, 
-    replace_vars, remove_dels, remove_dead, rename_labels)
- 
+from numba.ir_utils import (mk_unique_var, next_label, add_offset_to_labels,
+    replace_vars, remove_dels, remove_dead, rename_labels, find_topo_order, merge_adjacent_blocks)
+from numba.analysis import compute_cfg_from_blocks, find_top_level_loops
+from numba.transforms import _extract_loop_lifting_candidates, _loop_lift_get_candidate_infos
```
Comment (ninegua) [inline numba/inline_closurecall.py]: actually these imports are redundant, and should be removed.

File: numba/inline_closurecall.py [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
+from numba.targets.rangeobj import range_iter_len
+import numpy
```
Comment (ninegua) [inline numba/inline_closurecall.py]: This import is redundant and will be removed.

File: numba/inline_closurecall.py [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+
+enable_inline_arraycall = True
```
Comment (ninegua) [inline numba/inline_closurecall.py]: Yes, you are right. I'll put a comment.

File: numba/inline_closurecall.py [hunk 4]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+    pass
+
+def require(cond):
+    """
+    Raise GuardException if the given condition is False.
+    """
+    if not cond:
+       raise GuardException
+
+def guard(func, *args):
```
Comment (ninegua) [inline numba/inline_closurecall.py]: Not sure if `contextmanager` can help to achieve the same effect. The point of using `require` and `guard` (in many of the code that follows) is to refactor [nested if-statements](https://softwareengineering.stackexchange.com/questions/47789/how-would-you-refactor-nested-if-statements).

Typical solution (like mentioned in the above link) is using `if (!cond) { return None }`. But this has a few problems of its own: (1) each check becomes statement, and hence cannot be composed together like expressions; (2) the return value `None` still has to be manually checked by the caller; (3) `return None` doesn't play well with multiple return values. 

So the idiom of `require` and `guard` I use here is to convert between exceptions and optional values (with the presence of None). Similar tricks were done in functional programming languages like Haskell in the form of `Maybe` monad to ease programming in this kind of control flow. But Python doesn't have the same level of abstraction power, so I use exception to manipulate the control flow to achieve a similar effect.

Certainly I'm not the first one who invented this trick, I've seen it being used by other people. I wish I could give more pointers, but I don't exactly have them.

File: numba/inline_closurecall.py [hunk 5]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+            stmts.append(stmt)
+    debug_print("removed variables: ", removed)
+
+    # Define an index_var to index the array.
+    # If the range happens to be single step ranges like range(n), or range(m, n),
+    # then the index_var correlates to iterator index; otherwise we'll have to
+    # define a new counter.
+    range_def = guard(_find_iter_range, func_ir, iter_var)
+    index_var = scope.make_temp(loc)
+    index_offset_var = None
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will remove `index_offset_var`.

File: numba/inline_closurecall.py [hunk 6]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+        stmts.append(_new_definition(func_ir, index_var, ir.Const(value=-1, loc=loc), loc))
+
+    # Insert statement to get the size of the loop iterator
+    size_var = scope.make_temp(loc)
+    if range_def:
+        start, stop = range_def
+        # simple range with start and stop, which are defined only in caller,
+        # So we'll have to pass them as parameters
+        if start == 0:
+            # size_val = stop
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will remove the commented line.

File: numba/inline_closurecall.py [hunk 7]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+            size_val = stop
+        else:
+            # size_val = stop - start
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will remove the commented line since the next line is pretty much self-explanatory.

File: numba/inline_closurecall.py [hunk 8]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+            size_val = ir.Expr.binop(fn='-', lhs=stop, rhs=start, loc=loc)
+    else:
+        len_func_var = scope.make_temp(loc)
+        stmts.append(_new_definition(func_ir, len_func_var,
+                     ir.Global('range_iter_len', range_iter_len, loc=loc), loc))
+        # size_var = len_func(iter_var)
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will remove the commented line since the next line is pretty much self-explanatory.

File: numba/inline_closurecall.py [hunk 9]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+        size_val = ir.Expr.call(len_func_var, (iter_var,), (), loc=loc)
+
+    stmts.append(_new_definition(func_ir, size_var, size_val, loc))
+
+    size_tuple_var = scope.make_temp(loc)
+    # size_tuple_var = build_tuple(size_var)
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will remove the commented line since the next line is pretty much self-explanatory.

File: numba/inline_closurecall.py [hunk 10]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+    def find_nest_level(label):
+        level = 0
+        for loop in all_loops:
+            if label in loop.body:
+                level += 1
+    """
+
+    def find_array_def(arr):
+        """Find numpy array definition such as arr = numpy.empty(...).  If it is
+        arr = b[...], find array defintion of b recursively.
```
Comment (ninegua) [inline numba/inline_closurecall.py]: will correct the typo.

File: numba/inline_closurecall.py [hunk 11]
```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
...
+        arr_def = func_ir.get_definition(arr)
+        _make_debug_print("find_array_def")(arr, arr_def)
+        if isinstance(arr_def, ir.Expr):
+            if guard(_find_numpy_call, func_ir, arr_def) == 'empty':
+                return arr_def
+            elif arr_def.op == 'getitem':
+                return find_array_def(arr_def.value)
+        raise GuardException
+
+    def fix_array_assign(stmt):
```
Comment (ninegua) [inline numba/inline_closurecall.py]: The situation arises when there is nested comprehension as an argument to the array call.

File: numba/tests/test_comprehension.py [hunk 12]
```diff
@@ -251,6 +225,166 @@ def list24(x):
             msg = "Invalid usage of == with parameters"
             self.assertIn(msg, str(raises.exception))
 
+class TestArrayComprehension(unittest.TestCase):
+
+    @tag('important')
+    def test_comp_with_array_1(self):
```
Comment (ninegua) [inline numba/tests/test_comprehension.py]: Interesting. I think we have largely ignored complex numbers in testing, will investigate further.

Comment (ninegua) [inline numba/tests/test_comprehension.py]: The following code works:
```
a = np.empty(10, dtype=np.complex64)
a[0] = 1.j
```
However, the problem is how do we produce an IR for `empty(..., dtype=...)` before type inference? The information about element being `complex64` is only known after inference.

Similar situation is with list comprehension, where `build_list` comes before `list.append(...)`. However, for list this works out with no problem (before array-comprehension, or set `enable_inline_arraycall = False`):

```
l = [ 1j * x for x in range(10) ]
a = np.array(l)
```

So one way to solve this would require some tweaking in type inference to fill in the correct `dtype` data if it is missing from the original `array(..)` call. But this, however, is not quite the semantics of numpy. Any thoughts? @sklam

File: numba/inline_closurecall.py (hunk 2 above)
Comment (ninegua) [inline numba/inline_closurecall.py]: I looked at it again, `numpy` is only imported for its module name, not for accessing numpy functions. Do you still prefer using `import numpy as np` in this case?

File: numba/inline_closurecall.py (hunk 4 above)
Comment (stuartarchibald) [inline numba/inline_closurecall.py]: I have to admit this was an in part academic interest question, I had a hack at the code to try and do it with a ctx manager and it shorten/cleaned some parts but I got stuck (potentially irretrievably) in a couple of places, so was interested in your view. Require-guard as an idiom for this seems as good a way as any in Python. I imagine if there was an `Optional` type cf. Java 8's `java.util.Optional` that help in another way, but again, different language :)

Comment (stuartarchibald) [inline numba/inline_closurecall.py]: (and thanks for the thoughts/explanation/reasoning)

File: numba/tests/test_comprehension.py (hunk 12 above)
Comment (sklam) [inline numba/tests/test_comprehension.py]: I am looking into the this and experimenting with the use of the "refine_map" in the typeinfer to refine the array dtype just like how we did it with list.  Basic concept works but it is breaking a lot of tests.  I will update this when I make more progress.

Comment (ninegua) [inline numba/tests/test_comprehension.py]: Good to know. Thanks!

Comment (sklam) [inline numba/tests/test_comprehension.py]: I've made a PR with the dtype refinement on your branch.  https://github.com/IntelLabs/numba/pull/18

**Comment by sklam** [comment]
I just notice that the `dtype` argument is ignored when doing the following:

```python
In [1]: from numba import njit

In [2]: import numpy as np

In [3]: @njit
   ...: def foo(x):
   ...:     return np.array([i for i in range(x)], dtype=np.float32)
   ...: 

In [4]: foo(2)
Out[4]: array([ 0.,  1.])

In [5]: foo(2).dtype
Out[5]: dtype('float64')

In [6]: @njit
   ...: def foo(x):
   ...:     return np.array([i for i in range(x)], dtype=np.int32)
   ...: 
   ...: 

In [7]: foo(2)
Out[7]: array([ 0.,  1.])

In [8]: foo(2).dtype
Out[8]: dtype('float64')

```

**Comment by ninegua** [comment]
Commit 084e60f should address the explicit `dtype` argument to array call. The test `test_array_comp_with_dtype` covers this case.

**Comment by sklam** [comment]
The errors from https://travis-ci.org/numba/numba/builds/253297821 seems to only occur when the merged to master.

**Comment by sklam** [comment]
@ninegua, if you merge/rebase master and apply https://github.com/sklam/numba/commit/33e588bb11e115485fd3bdf30f0e0b5f747f60b4, it should fix it.

**Comment by seibert** [comment]
I think the merger of the random support PR has created a conflict.  Can you merge master into this PR again?

**Comment by ninegua** [comment]
@seibert I just did a rebase and pushed, and things should be good to go once Travis and AppVeyor check out.

File: numba/tests/test_comprehension.py [hunk 13]
```diff
@@ -248,14 +222,255 @@ def list24(x):
...
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+
+    @tag('important')
+    def test_comp_with_array_noinline(self):
+        def comp_with_array_noinline(n):
+            m = n * 2
+            l = np.array([i + m for i in range(n)])
+            return np.sum(l)
+        import numba.inline_closurecall as ic
+        ic.enable_inline_arraycall = False
```
Comment (stuartarchibald) [inline numba/tests/test_comprehension.py]: Should this have some sort of fail management? If an exception is raised in the following `ic.enabled_inline_arraycall = False` persists?

File: numba/unsafe/__init__.py [hunk 14]
```diff
@@ -0,0 +1,7 @@
+"""
+This subpackage is intented for low-level extension developers and compiler
+developers.  Regular user SHOULD NOT use code in this module.
+
+This contains compilable utility functions that can interacts directly with
```
Comment (stuartarchibald) [inline numba/unsafe/__init__.py]: s/interacts/interact/

File: numba/unsafe/ndarray.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
+"""
+This file provide internal compiler utilities that support certain special
```
Comment (stuartarchibald) [inline numba/unsafe/ndarray.py]: s/provide/provides/

File: numba/unsafe/ndarray.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+
+
+@intrinsic
+def empty_inferred(typingctx, shape):
+    """A version of numpy.empty whose dtype is inferred by the type system.
+
+    Expects `shape` to be a int-tuple.
+
+    There are special logic in the type-inferencer to handle the "refine"-ing
```
Comment (stuartarchibald) [inline numba/unsafe/ndarray.py]: s/are/is/

File: docs/source/reference/pysupported.rst [hunk 17]
```diff
@@ -138,6 +138,24 @@ List comprehension
...
+    def f(n):
+      return np.array([ [ x * y for x in range(n) ] for y in range(n) ])
+
+In this case, Numba is able to optimize the program to allocate and 
+initialize the result array directly without allocating intermediate 
+list objects.  Therefore, the nesting of list comprehension here is 
+not a problem since no nested list is being created.
+
+Additionally, Numba supports parallel array comphension when combined 
+with the :ref:`parallel_jit_option` option on CPUs.
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]: Is the functionality provided by `fix_array_assign` worth a mention, perhaps that would be better placed in developer docs?

**Comment by stuartarchibald** [comment]
Also, @seibert CI may need some manual encouragement please as ef227cb doesn't appear to have run?

File: numba/inline_closurecall.py [hunk 18]
```diff
@@ -236,19 +281,481 @@ def _replace_returns(blocks, target, return_label):
...
+        extra_dims = [ _get_definition(func_ir, x, lhs_only=True) for x in dim_def.items ]
+        debug_print("extra_dims = ", extra_dims)
+        # Expand size tuple when creating lhs_def with extra_dims
+        size_tuple_def = _get_definition(func_ir, lhs_def.args[0])
+        require(isinstance(size_tuple_def, ir.Expr) and size_tuple_def.op == 'build_tuple')
+        debug_print("size_tuple_def = ", size_tuple_def)
+        size_tuple_def.items += extra_dims
+        # In-place modify rhs_def to be getitem
+        rhs_def.op = 'getitem'
+        rhs_def.value = _get_definition(func_ir, lhs, lhs_only=True)
```
Comment (ninegua) [inline numba/inline_closurecall.py]: The call to `_get_definition` will not raise because a previous `find_array_def(lhs)` (which calls _get_definition, but without setting `lhs_only`) have already returned a value.

File: numba/parfor.py [hunk 19]
```diff
@@ -1873,6 +1888,37 @@ def get_parfor_array_accesses(parfor, accesses=None):
 ir_utils.array_accesses_extensions[Parfor] = get_parfor_array_accesses
 
 
+def parfor_add_offset_to_labels(parfor, offset):
+    #print("parfor_add_offset_to_labels")
```
Comment (ninegua) [inline numba/parfor.py]: Will remove them.

File: docs/source/reference/pysupported.rst (hunk 17 above)
Comment (ninegua) [inline docs/source/reference/pysupported.rst]: The `fix_array_assign` trick was to support mutli-dimension array when there is nested comprehension, which actually was shown in the quoted example. I'll reword it to better deliver the message that nested comprehension creates multi-dimensional array, and is supported.

File: numba/unsafe/ndarray.py (hunk 15 above)
Comment (ninegua) [inline numba/unsafe/ndarray.py]: will fix.

File: numba/unsafe/ndarray.py (hunk 16 above)
Comment (ninegua) [inline numba/unsafe/ndarray.py]: will fix.

File: numba/unsafe/__init__.py (hunk 14 above)
Comment (ninegua) [inline numba/unsafe/__init__.py]: will fix.

File: numba/inline_closurecall.py (hunk 18 above)
Comment (sklam) [inline numba/inline_closurecall.py]: I've made a patch (https://gist.github.com/sklam/0170e573b5b08272f35a31c2b57a3091) to avoid the inplace mutation of the `rhs_def` and the deleting of internal attrs.  The patch splits the inplace operation into a mark phase and replace phase.  Do you think it is valid?

File: numba/tests/test_comprehension.py (hunk 13 above)
Comment (ninegua) [inline numba/tests/test_comprehension.py]: Good point! I will fix it.

File: numba/tests/test_comprehension.py [hunk 20]
```diff
@@ -248,14 +222,255 @@ def list24(x):
...
+
+        cfunc = jit(nopython=True)(comp_with_array_1)
+        np.testing.assert_array_equal(comp_with_array_1(5), cfunc(5))
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+
+        # test parallel array comprehension
+        cfunc = jit(nopython=True, parallel=True)(comp_with_array_1)
+        np.testing.assert_array_equal(comp_with_array_1(5), cfunc(5))
+        self.assertNotIn('allocate list', cfunc.inspect_llvm(cfunc.signatures[0]))
+        self.assertIn('@do_scheduling', cfunc.inspect_llvm(cfunc.signatures[0]))
```
Comment (ninegua) [inline numba/tests/test_comprehension.py]: Good suggestion! Will refactor the code.

File: numba/tests/test_comprehension.py [hunk 21]
```diff
@@ -248,14 +222,219 @@ def list24(x):
...
         if utils.PYVERSION < (3, 0):
             with self.assertRaises(TypingError) as raises:
                 cfunc = jit(nopython=True)(list22)
                 cfunc(var)
             msg = "cannot unify reflected list(int%d) and int%d" % (bits, bits)
             self.assertIn(msg, str(raises.exception))
 
+class TestArrayComprehension(unittest.TestCase):
+
+    def check(self, pyfunc, *args, run_parallel=False, assert_allocate_list=False,
```
Comment (seibert) [inline numba/tests/test_comprehension.py]: This function signature does not work in Python 2.7.  `*args` can't be followed by explicit keyword args, only `**kwargs`.  Since this is just test code, it might be easier to pass the extra positional args as a list.

**Comment by seibert** [comment]
The Python 2.7 tests are failing because of the syntax error I noted above.

File: numba/inline_closurecall.py (hunk 18 above)
Comment (ninegua) [inline numba/inline_closurecall.py]: @sklam The patch does not work because one replacement must be completed before the next replacement can be considered, which is the case for nested array comprehension. 

That said, the original inplace update does not enforce a block traversal order, which is now fixed in commit abc5c78324fc7842534d4e141f16d2b90531833c

File: numba/tests/test_comprehension.py (hunk 21 above)
Comment (ninegua) [inline numba/tests/test_comprehension.py]: Ok, will do.

**Comment by seibert** [comment]
OK, I think this is ready to merge.   Thanks for this contribution!
